### PR TITLE
perf(decode): fused sequence executor + SIMD/FSE hot-path cleanup + DoS-safe rollback

### LINKED ISSUE #178: perf(decode + encode-greedy): close 3-5× donor gap on negative-level decompress; share SIMD primitives + add dedicated greedy strategy
## Status (re-measured)

The original 3-5× negative-level decode gap is closed to **1.3-1.4×**; the missing dedicated greedy strategy has landed. This issue now tracks the residual negative-level decode parity work only.

Current numbers (i9, decodecorpus-z000033, level_-1_fast):

| Stream source | Rust decode | C decode | Gap (was) |
|---------------|-------------|----------|-----------|
| `c_stream` | 907 µs (1.05 GiB/s) | 647 µs (1.47 GiB/s) | **1.40×** (was 3.69×) |
| `rust_stream` | 1.18 ms (824 MiB/s) | 895 µs (1.06 GiB/s) | **1.32×** (was 5.35×) |

## Landed since the original report

- **Greedy strategy**: dedicated `StrategyTag::Greedy` (L5) with `lazy_depth = 0` on the Row finder — the reference's own greedy shape (its greedy/lazy share the row-search template with depth 0). The L4 dfast outlier was separately closed by the donor greedy double-fast port.
- **XXH64 frame checksum**: gated on the frame's checksum flag (−61% on flag-off frames) and hashed per-block while cache-hot on the direct path; no longer a post-decode cold walk.
- **Per-kernel decoders**: four ISA tiers (Scalar/BMI2/AVX2/VBMI2) with monolithic per-tier sequence loops, BMI2 bit-reader specialization, per-tier match-copy chains.
- **SIMD wildcopy / overshooting copies**: donor-shape overshoot-tolerant copies with bounded tails per kernel.

## Residual (the 1.3-1.4×)

The remaining gap is the known sequence-decode body delta vs the reference's single `bmi2.constprop` monolith (HUF 4-stream burst on literal-heavy frames being the biggest single item on weak-compression fixtures). Tracked levers, in ROI order:

1. HUF burst decompress port (one inlined monolith vs our 3-fn split) — order-of-magnitude self-time delta on literal-heavy frames.
2. Sequence-loop body instruction diff vs the reference per-tier.

## Kill-switch criteria

Stop pulling individual levers when both stream sources sit within ~1.1× of the reference on the negative-level corpus, or when a lever returns <2% twice in a row (record the negative result and move on).

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Continuation of decode-speed work that was on PR #190 (closed because that PR's history accumulated a panic-guard variant of the malformed-input fix that a libFuzzer artifact (`crash-bfb3bc55…`) then turned into a DoS surface). This PR ships the same decode-speed improvements **plus** the graceful-rollback fix that closes that surface.

Closes #178

## Decode speed

Baseline (z000033, Level 3 dfast, M1): 1,814,570 ns/iter → best 1,353,635 ns / median ~1,413,000 ns (≈ **−22%**). Throughput 553 → 739 MB/s. FFI gap 3.89× → 2.88×.

## Notable changes

- **Fused sequence executor** (`decoding/sequence_section_decoder.rs`): collapses `decode_sequences` + `execute_sequences` into a single loop, eliminating the `Vec<Sequence>` roundtrip on the hot path.
- **Transactional rollback** for the fused executor via `DecodeBuffer::checkpoint` / `try_restore_checkpoint`. The rollback primitive snapshots the `RingBuffer` `cap` alongside `tail` so an intervening reallocation (only reachable on malformed input that decodes past `MAX_BLOCK_SIZE`) is detected and the rollback is gracefully skipped, **not** turned into a panic. The caller still surfaces a normal decode `Err`.
- **SIMD copy** dispatcher collapsed; single 16-byte fast path; exact-tail path widened to 32 B (4× `u64`); AVX2 / AVX-512 kernels target-feature-gated for `no_std`.
- **BitReaderReversed::refill** split into hot inline + `#[cold] #[inline(never)]` slow path; `mask_lower_bits` computed via shift instead of `BIT_MASK[n]` LUT.
- **RingBuffer** `head + start == cap` now routes through the contiguous-second-slice path (was misrouted into split-source, producing a zero-length `copy_bytes_overshooting` against the slack region).
- **`do_offset_history`** fast path for `offset_value >= 4`, matching donor `ZSTD_updateRep`.
- **no-std CI gate** + clippy on `--no-default-features` and `--no-default-features --features hash`.

## Tests

- 506/506 lib tests pass (505 + new `malformed_block_does_not_panic_via_restore_checkpoint` fuzz regression).
- Fuzz artifact `crash-bfb3bc55…` (the input that took down PR #190's CI) is now an inline regression test in `zstd/src/tests/fuzz_regressions.rs`.
- All clippy modes (`--all-targets`, `--no-default-features`, `--no-default-features --features hash`) clean.

## Related

- Replaces PR #190 (closed).

## Summary by CodeRabbit

* **New Features**
  * Added CI checks for no-standard-library builds to catch regressions on minimal targets.

* **Bug Fixes**
  * Safer rollback/checkpoint restore to avoid panics on malformed input and prevent buffer clobbering.
  * Improved edge-case decoding robustness and validated offset/history handling.

* **Tests**
  * Added regression tests for malformed frames and capacity-edge copy behavior.

* **Chores**
  * Performance and SIMD-backed copy-path optimizations for decompression.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/194?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: zstd/src/decoding/sequence_section_decoder.rs [hunk 1]
```diff
@@ -35,14 +57,190 @@ pub fn decode_sequences(
...
+        // try_restore_checkpoint succeeds when no reallocation happened
+        // between the checkpoint and now (the common case: upfront
+        // reserve(MAX_BLOCK_SIZE) covers a well-formed block). When a
+        // malformed block decodes past that bound, reserve_amortized
+        // fires and compacts the ring buffer — the captured tail is no
+        // longer meaningful and the rollback is skipped. Either way the
+        // caller observes the same Err below; the partial data left in
+        // the buffer in the latter case is discarded with the frame.
+        let _restored = buffer.try_restore_checkpoint(buffer_checkpoint);
+        *offset_hist = saved_offset_hist;
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/decoding/sequence_section_decoder.rs]: @coderabbitai Fixed in 5478b51f — offset_hist rewind now gated on try_restore_checkpoint() returning true. Also addressed both outside-diff findings: update_state_fast → pub(crate); ringbuffer::reserve uses saturating_sub on the fast path + checked_add in reserve_amortized.
